Source/WTF/wtf/NativePromise.h

```diff
@@ -340,7 +342,7 @@ class NativePromise final : public NativePromiseBase, public ConvertibleToNative
         Locker lock { m_lock };
         PROMISE_LOG(resolveSite, " resolving ", *this);
         ASSERT(isNothing());
-        m_result = std::forward<ResolveValueType_>(resolveValue);
+        m_result = crossThreadCopy(std::forward<ResolveValueType_>(resolveValue));
```

Comment:
This is good to cross thread copy since NativePromise may be destroyed in any thread.
But I am not sure this is enough, we probably want to ensure that m_result is cross thread copied for every then callback, since each one of them may use m_result on a different queue.
In that sense, I would remove NativePromise::result().

If we add that, it would be nice to keep the optimisation for the one then callback native promise.

Reply:
No, it doesn't have to be.

If IsExclusive==true, it is moved.

If IsExclusive==false : the object is passed via const reference; and the object is guaranteed to exist and not be modified while the callback is being called. Even if accessed from a different queue, an originally isolated object is usable (at least for all the objects we currently have in the tree).

As such, reading (or copying) the result is safe, regardless of the value of IsExclusive.

That only thing that isn't is taking a refcount of a non thread-safe refcounted value; but for that there's not much we can do as and is already protected through runtime assertions

Reply:
> But I am not sure this is enough, we probably want to ensure that m_result is cross thread copied for every then callback, since each one of them may use m_result on a different queue. In that sense, I would remove NativePromise::result().

Ok.

If I fully understand your aim, you're hoping that for object that have non-thread safe  refcounting you would be able to use them safely across multiple-threads through some hidden, under the hood mechanism.

And that is not achievable because at the time you resolve or reject a promise, it may not have any ThenCommand attached yet. A NativePromise can be then()ed on any threads. 
At the time you then()ed you may not be on the producer's thread anymore, you've lost the ability to safely perform another ref() 

Consider the following:
```
Ref<NonExclusiveNativePromise> p = promiseProducer(); /* returned promise has already been resolved */
m_workQueue->dispatch[p = WTFMove(p)] {
  p->whenSettled(RunLoop::main(), [](const NonExclusiveNativePromise::Result& result) { /* do something else */ }
}
```

As it is, the reference that the callback receives to the resolve/reject object is now immutable and read-only. You can perform a copy, but you can't take an extra ref() to it in a safe fashion. If you need to achieve this, then the object needs to have thread-safe refcounting. There's no other way about it.

We can store a `crossThreadCopy` at the time resolve/reject and pass that around, which is what this PR is doing.

This can be explained and documented, with the caveats listed, but C++ isn't Rust unfortunately, there are restrictions.

Reply:
As a follow-up patch, I can make the use of IsExclusive=false forbidden if the Resolve/RejectValueT contains a String or AtomString with a static_assert.

This will cover your concerns about misuse of the promise.

Reply:
I have implemented a new strategy with this latest PR.

You now define the behaviour with a PromiseOption enum.
There are two relevants:
WithCrossThreadCopy and WithoutCrossThreadCopy, if neither are set (the default) then it will automatically use crossThreadCopy on the resolve/reject value if one of the type used contains an AtomString or a method `isolatedCopy` 
Creating a non-exclusive NativePromise without setting either WithCrossThreadCopy or WithoutCrossThreadCopy is now forbidden.
if `WithCrossThreadCopy` is set, it will be up to the consumer to call crossThreadCopy() on the resolve/reject value provided as needed.